### ISSUE TITLE
Ubiquity Banking On-Ramp Experience (*.eth -> UUSD.gno)

### ISSUE THREAD

**Issue by 0x4007** [ubiquity/uusd.ubq.fi#4]
> [!CAUTION]
> Make all pages separate HTML pages. Do not page a single page application. This makes maintenance hell. 

### Objective

Currently we have some hoops to jump through to mint UUSD and to use on Gnosis Chain: 
- acquire 95% LUSD
- acquire 5% UBQ
- mint UUSD
- bridge to Gnosis Chain

It would be ideal to abstract this all away from the end user, gas fees and all. 

The solution would require a combination of smart contracts and UI magic. 

### Minting

We can brand it as something that implies a slower process instead of just "instant minting" perhaps a term from traditional finance like "on-ramp to UbiquityOS" or like a "bank transfer" which typically takes a business day (ours I expect to take like 30 minutes.)

When a user starts with any crypto asset in their wallet, like ETH, and wants to mint UUSD to use on Gnosis Chain (starting from mainnet) we need to do the following:

1. Cowswap order for 95% LUSD, 5% UBQ
2. Transfer the UUSD to Gnosis Bridge
3. Notify the user that their "Ubiquity Bank Account Balance" will be updated in approximately 30 minutes. 

Within the price quote before they begin the process, we need to clearly show in the modal their starting value and, add a clear line item for "processing fees" and show their final estimated "bank balance" 

The on-ramping experience makes sense for corporations using UbiquityOS to pay their teams. It would be nice to think of reasons for contributors to use this flow as well. Perhaps to top up their Ubiquity Cards in the future when we support real debit cards. 

### Implementation Steps
- [ ] Create MVP UI
- [ ] Calculate "Processing Fees"

### Remarks

Core ethos of Ubiquity is to try and abstract as much of the nitty-gritty/tech from the end users. In some years people won't care about "blockchain" and "AI" instead they care about a wonderful end user experience that solves real problems. 

Bear this in mind when implementing.

**Comment by 0x4007** [comment]
@rndquu rfc

**Comment by rndquu** [comment]
> @rndquu rfc

So the main issue is that right now it takes too many steps to bridge UUSD from mainnet to gnosis.

We could add a new UI page similar to:
<img width="860" alt="Screenshot 2024-10-01 at 13 17 56" src="https://github.com/user-attachments/assets/efccc187-c0c8-45ff-b398-1e6d64fabb99">

How it could work:
1. User inputs a token amount (`0.1 ETH` or `100 USDT`)
2. User clicks the "Bridge" button
3. [Multicall](https://github.com/Uniswap/v3-periphery/blob/0682387198a24c7cd63566a2c58398533860a5d1/contracts/base/Multicall.sol) tx is created which:
  a) Swaps `0.1 ETH` for `LUSD` and `UBQ`
  b) Mints `UUSD`
  c) Initializes a bridge (we should check how contracts work at https://bridge.gnosischain.com/)

**Comment by 0x4007** [comment]
Might be relevant, was using GitHub search:
- https://github.com/gnosischain/tokenbridge
- https://github.com/gnosischain/omnibridge

**Comment by rndquu** [comment]
> Might be relevant, was using GitHub search:
> 
> * https://github.com/gnosischain/tokenbridge
> * https://github.com/gnosischain/omnibridge
> 
> You think you can help me break down a checklist of <1 Week tasks, or verify my existing implementation steps so we can fund them?

Implementation of these steps should take 1-3 days:
1. User inputs a token amount (0.1 ETH or 100 USDT)
2. User clicks the "Bridge" button
3. [Multicall](https://github.com/Uniswap/v3-periphery/blob/0682387198a24c7cd63566a2c58398533860a5d1/contracts/base/Multicall.sol) tx is created which:
  a) Swaps 0.1 ETH for LUSD and UBQ
  b) Mints UUSD
  c) Initializes a bridge (we should check how contracts work at https://bridge.gnosischain.com/)

There is no need for any custom smart contracts in the flow above.

**Comment by zugdev** [comment]
@0x4007 by abstracting gas costs you mean taking input amount as a transfer and include swap and bridge gas costs as processing fees correct?

**Comment by 0x4007** [comment]
Yes the fees come out of the user's balance they are trying to onramp with

**Comment by zugdev** [comment]
Great, am I allowed to go with React? Very easy to deploy, fairly easy to maintain since it's modular. Also it's pretty popular and there will always be people who know it to code new features / do maintenance.

**Comment by zugdev** [comment]
By "Make all pages separate HTML pages" you mean also separating CSS right? I am also not very sure what multiple pages this app would have. But if there is need for multiple feature/pages I'll surely split it.

**Comment by zugdev** [comment]
Also is there a preference for wallet provider? I usually go with rainbow kit + wagmi. However if there is a specific preference let me know.

**Comment by zugdev** [comment]
We could in future use an aggregator pre-step to allow minting UUSD to Gnosis with assets from any chain too.

**Comment by 0x4007** [comment]
> Also is there a preference for wallet provider? I usually go with rainbow kit + wagmi. However if there is a specific preference let me know.

I think wallet connect or whatever they rebranded to @rndquu might have some thoughts on this

**Comment by rndquu** [comment]
> > Also is there a preference for wallet provider? I usually go with rainbow kit + wagmi. However if there is a specific preference let me know.
> 
> I think wallet connect or whatever they rebranded to @rndquu might have some thoughts on this

For https://uad.ubq.fi/ we used Wallet Connect (it seems it was rebranded to https://reown.com/) + for all of our projects we use https://ethers.org/ for blockchain related tasks.

`rainbow kit + wagmi` are modern tools while `Wallet Connect + ethers.js` is used everywhere in Ubiquity repos, so it's a tradeoff between modern tooling and consistency across all repos.

**Comment by zugdev** [comment]
We can go WC + ethers no problem

**Comment by zugdev** [comment]
Can someone explain where the 5% UBQ will go? Are they burnt, paid in fees, or do they stay in treasury? I need to know so I can do math on the expected end amount of UUSD. For example, if I put in 100 DAI and we have perfect swap liquidity, so I can get 95 DAI worth of LUSD and 5 DAI worth of UBQ, after minting will my end amount of UUSD be 95 DAI worth of UUSD which is equal to the LUSD amount or is it the summation?

**Comment by 0x4007** [comment]
Burnt but I have always been hesitant on destroying the supply. Maybe we can change one day if there's a superior choice from an economics standpoint.

**Comment by zugdev** [comment]
Yes, not only economics but burning governance needs some thinking.

**Comment by zugdev** [comment]
Shouldn't we bump 'esbuild'? I am getting a lot of "Could not resolve 'dependency'" from inside dependencies. I really don't want to install `react` if we are not using React. 

![image](https://github.com/user-attachments/assets/31ecc66c-303c-4251-9340-ea1606649983)

Is there a way to navigate through this?

**Comment by zugdev** [comment]
@rndquu what is current UUSD address? The latest one I found, from transactions with the pool was: `0xb6919ef2ee4afc163bc954c5678e2bb570c2d103` but that symbol is uAD.

**Comment by 0x4007** [comment]
Upgrading esbuild seems like it would not be a problem. Which has a react dependency, the wallet stuff i presume? 

If its the only clear path forward I guess you can install it as a dev dependency.

**Comment by zugdev** [comment]
Yes, it's reown (Wallet Connect).

**Comment by zugdev** [comment]
It will take more than a week certainly. Integrating with CoW Swap's APIs demands a lot of testing, handle caching, integrating bridge... I rather take longer and keep code quality up, that's a definer of maintainability.

**Comment by zugdev** [comment]
I probably have to setup oracles to estimate UUSD amount as well, if user inputs ETH I can only get the swap fees in ETH, therefore I need to estimate ETH price to calculate final UUSD amount.

**Comment by 0x4007** [comment]
I think it is worth you filing new child issues with smaller time frames if you think this will take more than a week and then you can get @rndquu approval on funding them.

**Comment by zugdev** [comment]
I am creating issues on the "hardest" sub-issues within this issue. The big problem here is that the multicall approach does not work: swaps are not instantaneously responded and they can fail. So are bridges. It also masks some big issues:

1. To abstract the costs and fees from the user we must have a back-end private key which will receive funds and then handle signing in all future steps. What if one step fails, do we rollback and send remaining funds back to user? We must prevent failure, or at least warn the user in likely scenarios, by perfectly managing slippage and quoting from each step.

2. To calculate the UUSD amount we must figure figure dollar price from input token, understand dollar price from all fees and also account for slippage.

3. Setting up a bridge is not a ridiculous task, I need to look into how existing mainnet -> gnosis bridges work, will any existing deployed bridge accept bridging our token? Will they want fees in another currency? If so we need to account for that in the swap phase as well, so we need retroactive fee calculation. Might even need to deploy proprietary bridge. Also state-of-the-art bridges allow users to get some gas tokens on the other side when they bridge, need to work on that too.

4. We will be integrating from 3-5 APIs we need to manage caching and limit rates very carefully. 

5. We need a storage and cache system for on-going tasks in the system, we don't want any task lost if server resets. This has to be as secure and atomic as a banking system or else funds will need to be manually returned all the time.

There are even more reasons but to sum up, yes a UI could be sketched in 3 days, could even set up swap -> mint -> bridge in close to 5 days, but it wouldn't be safe, nor fail-proof nor efficient. I'd rather make it fully good in one go than deliver a little piece of it to open a new issue and then continue developing, this way we can set a fine modular base to iterate on, handling all edge cases. I honestly think that to deliver a good UX solution here it would take me a month, I could have something working in less to facilitate team UUSD minting but to make it good it's more than that. 

This idea is really not that different from CoW Swap's flow, a swap like UI with fees calculation, fiat amount calculation, bridges embedeed... if you check their [UI repo](https://github.com/cowprotocol/cowswap), it's very big.

**Comment by 0x4007** [comment]
@rndquu rfc

**Comment by zugdev** [comment]
I think another task necessary would be to deploy a bridge or use Wormhole/Portal but I don't know if that's this repo's scope.

**Comment by 0x4007** [comment]
Use gnosis bridge

**Comment by gentlementlegen** [comment]
```diff
# Passed the deadline and no activity is detected, removing assignees: [@zugdev].
```

**Comment by rndquu** [comment]
@zugdev 

> Can someone explain where the 5% UBQ will go?

On [minting](https://github.com/ubiquity/ubiquity-dollar/blob/20799bedf5ee4d40d2863653f7bc2fe68a2ce3fa/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L543) UUSD for LUSD+UBQ: UBQ tokens are burnt
On [redeeming](https://github.com/ubiquity/ubiquity-dollar/blob/20799bedf5ee4d40d2863653f7bc2fe68a2ce3fa/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L651) LUSD+UBQ for UUSD: UBQ tokens are minted

> We need a crypto price API to calculate USD values

Take any free provider

**Comment by rndquu** [comment]
> I am creating issues on the "hardest" sub-issues within this issue. The big problem here is that the multicall approach does not work: swaps are not instantaneously responded and they can fail. So are bridges. It also masks some big issues:
> 
> 1. To abstract the costs and fees from the user we must have a back-end private key which will receive funds and then handle signing in all future steps. What if one step fails, do we rollback and send remaining funds back to user? We must prevent failure, or at least warn the user in likely scenarios, by perfectly managing slippage and quoting from each step.
> 2. To calculate the UUSD amount we must figure figure dollar price from input token, understand dollar price from all fees and also account for slippage.
> 3. Setting up a bridge is not a ridiculous task, I need to look into how existing mainnet -> gnosis bridges work, will any existing deployed bridge accept bridging our token? Will they want fees in another currency? If so we need to account for that in the swap phase as well, so we need retroactive fee calculation. Might even need to deploy proprietary bridge. Also state-of-the-art bridges allow users to get some gas tokens on the other side when they bridge, need to work on that too.
> 4. We will be integrating from 3-5 APIs we need to manage caching and limit rates very carefully.
> 5. We need a storage and cache system for on-going tasks in the system, we don't want any task lost if server resets. This has to be as secure and atomic as a banking system or else funds will need to be manually returned all the time.
> 
> There are even more reasons but to sum up, yes a UI could be sketched in 3 days, could even set up swap -> mint -> bridge in close to 5 days, but it wouldn't be safe, nor fail-proof nor efficient. I'd rather make it fully good in one go than deliver a little piece of it to open a new issue and then continue developing, this way we can set a fine modular base to iterate on, handling all edge cases. I honestly think that to deliver a good UX solution here it would take closer to a month, I could have something working in less to facilitate team UUSD minting but not for prod.
> 
> This idea is really not that different from CoW Swap's flow, a swap like UI with fees calculation, fiat amount calculation, bridges embedeed... if you check their [UI repo](https://github.com/cowprotocol/cowswap), it's very big.

I feel like this is 10x overengineered compared with what we really need.

> The big problem here is that the multicall approach does not work: swaps are not instantaneously responded

Check this example:
1. User inputs `0.1 ETH`
2. `uusd.ubq.fi` fetches `ETH/USD` quote on-chain from `ETH/USD` chainlink oracle
3. `uusd.ubq.fi` [fetches](https://github.com/ubiquity/ubiquity-dollar/blob/66899068ea4604354f9f0becbd7e10855d974cbc/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L433) `UBQ/USD` quote on-chain
4. `uusd.ubq.fi` fetches `LUSD/USD` quote on-chain from `LUSD/USD` chainlink oracle
5. Add tx to multicall: swap 95$ worth of ETH for LUSD + 5$ worth of ETH for UBQ (uniswap, pancakeswap, etc...)
6. Add tx to multicall: mint UUSD for LUSD+UBQ
7. Add tx to multicall: [bridge](https://bridge.gnosischain.com/) UUSD from mainnet to gnosis (as far as I understand [relayTokens](https://github.com/gnosischain/omnibridge/blob/c814f686487c50462b132b9691fd77cc2de237d3/contracts/upgradeable_contracts/components/common/TokensRelayer.sol#L54) method is used)
8. Send multicall tx

Why can't we use the flow described above?

**Comment by rndquu** [comment]
> @rndquu `CoW Swap` operates in a request/response swap (pull model), a swap transaction is actually an order: you have to wait for it to be fulfilled and it's not even deterministic.

We don't need `CoW Swap` for v1, `pancake` or any popular DEX is enough. Right now the main issue is that it takes too many steps for `0x4007` to convert mainnet `DAI` to gnosis `UUSD` while it could be done in a single UI page in a single transaction.

**Comment by zugdev** [comment]
I dont think we can do it in a single tx, we gotta swap then mint then bridge, not swap and mint and bridge at the same time

**Comment by rndquu** [comment]
> Also, please remember multicall is just a transaction bundler, it is not synchronous nor sequential

It is synchronous and sequential

**Comment by zugdev** [comment]
Ok not using CoW Swap makes swapping a one block process. But initiating a bridge in block 1 does not mean funds are available in block 2. However, solving the swap is already great, we can initiate bridge with user as receiver and might work. I will try that!
